### PR TITLE
make MIR graphviz generation use gsgdt

### PR DESCRIPTION
**Pull request by vn-ki**
gsgdt [https://crates.io/crates/gsgdt] is a crate which provides an
interface for stringly typed graphs. It also provides generation of
graphviz dot format from said graph.

This is the first in a series of PRs on moving graphviz code out of rustc into normal crates and then implementating graph diffing on top of these crates.

r? @oli-obk

### CODE REVIEW

**Review by oli-obk — COMMENTED** [review]
cc @rust-lang/wg-mir-opt

**Comment by richkadel** [comment]
@vn-ki - I just saw this PR was mentioned after bors landed a conflicting commit from my PR #78267 . I like your goals of moving graphviz code out of rustc. Note that my PR added a [generic_graphviz.rs](https://github.com/rust-lang/rust/blob/master/compiler/rustc_mir/src/util/generic_graphviz.rs) implementation, based mostly on the MIR graphviz.rs in the same directory. I needed a MIR-like graphviz output for coverage graphs, but didn't want to create another type-specific 1-off.

You may need to look that over and decide how to work in your changes, whether replacing my implementation or not. FYI, I implemented enough hooks (by generic handler functions) to support the coverage graph, but not all of the hooks required to completely implement an equivalent to the existing MIR graphviz.rs implementation. It wouldn't be hard to extend generic_graphviz.rs to do that though. But you may or may not want to consider that approach. (If this is even relevant to your change.)

**Comment by vn-ki** [comment]
This PR does not intend to touch coverage graphs, so I guess both implementations will be in rustc (for now). 

I will move the rest of the graphviz consumers to gsgdt in future PRs, and then we will be able to finally remove all graphviz code. 

Most of the current logic in `mir::util::graphviz` has been ported to gsgdt because I thought that would be a general approach that will be required for most graphs of such nature. So I think we won't be needing generic_graphviz, though I'm not sure.

File: Cargo.lock [hunk 1]
```diff
@@ -5252,7 +5262,7 @@ dependencies = [
  "chrono",
  "lazy_static",
  "matchers",
- "parking_lot 0.9.0",
+ "parking_lot 0.11.0",
```
Comment (vn-ki) [inline Cargo.lock:5265]: Not sure what caused this. Some rebase magic. Should I revert this change?

Comment (oli-obk) [reply Cargo.lock:5265]: yea, please revert it

File: compiler/rustc_mir/src/util/generic_graph.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    Graph::new(graph_name, nodes, edges)
+}
+
+fn bb_to_graph_node(block: BasicBlock, body: &Body<'_>, dark_mode: bool) -> Node {
+    let def_id = body.source.def_id();
+    let data = &body[block];
+    let label = node(def_id, block);
+
+    let (title, bgcolor) = if data.is_cleanup {
+        (format!("{} (cleanup)", block.index()), "lightblue")
```
Comment (richkadel) [inline compiler/rustc_mir/src/util/generic_graph.rs:44]: I just noticed you missed a bug fix during the rebase conflict resolution:

This should be:

```rust
        let color = if dark_mode { "royalblue" } else { "lightblue" };
        (format!("{} (cleanup)", block.index()), color)
```

See line 116 in the deleted sections of graphviz.rs

**Review by richkadel — CHANGES_REQUESTED** [review]

**Comment by oli-obk** [comment]
@bors retry

```
    Updating crates.io index
warning: spurious network error (2 tries remaining): error inflating zlib stream; class=Zlib (5)
warning: spurious network error (1 tries remaining): error inflating zlib stream; class=Zlib (5)
error: failed to get `cc` as a dependency of package `bootstrap v0.0.0 (D:\a\rust\rust\src\bootstrap)`
```

**Comment by oli-obk** [comment]
Oh... so... this is a problem we usually just had when rebuilding clippy within the worktree. But now that `serde_derive` is also built during compilation of rustc, this will happen deterministically.

cc @rust-lang/clippy maybe we should revisit our test suite? I opened a zulip thread about it in https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/multiple.20matching.20crates/near/216590414

**Comment by camelid** [comment]
This has the `S-waiting-on-fcp` label -- is it still waiting on one?

**Comment by jyn514** [comment]
I think you could try loading `serde_derive` from the sysroot instead of building from cargo. The error is that it's in both sysroot and in the `--extern` directory passed by cargo:

```
error[E0464]: multiple matching crates for `serde_derive`
  --> $DIR/used_underscore_binding_macro.rs:4:1
   |
LL | extern crate serde_derive;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: candidates:
           crate `serde_derive`: /checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps/libserde_derive-8823e1c401e486b.so
           crate `serde_derive`: /checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libserde_derive-4b65394b709f9e2e.so
```
It will be in the sysroot whether clippy is in-tree or out of tree, since `rustc_private` gives you access to *all* crates in the sysroot.

**Comment by bjorn3** [comment]
The `--extern` should be favored over the sysroot.

**Comment by oli-obk** [comment]
@bors r-

what happened here?

**Comment by camelid** [comment]
@rust-lang/infra bors seems to have tried to merge this PR even though it shouldn't have been in the approved state. That, or someone `r=`'d and then deleted their comment?

**Comment by camelid** [comment]
Also, Oli's `@bors r-` didn't seem to work.

**Comment by camelid** [comment]
Should I have cc'd Pietro instead? I know they work on bors, and they don't seem to be on the infra GH team...

**Comment by Mark-Simulacrum** [comment]
Pietro is absolutely on the infra GH team and you should always ping the team. I think once bors starts testing you need to r- retry.

**Comment by camelid** [comment]
> Pietro is absolutely on the infra GH team and you should always ping the team.

Ah, GitHub was only showing three team members in the hovercard, so it confused me.

> I think once bors starts testing you need to r- retry.

Hmm, maybe that should be changed so just an `r-` works? It seems like somewhat unintuitive behavior.

**Comment by ebroto** [comment]
I also hit this.

TL;DR I think you can hotfix this by adapting the test, the changes should not affect what is being tested.
```diff
diff --git a/src/tools/clippy/tests/ui/crashes/used_underscore_binding_macro.rs b/src/tools/clippy/tests/ui/crashes/used_underscore_binding_macro.rs
index 6d2124c12fe..c57a45dc7aa 100644
--- a/src/tools/clippy/tests/ui/crashes/used_underscore_binding_macro.rs
+++ b/src/tools/clippy/tests/ui/crashes/used_underscore_binding_macro.rs
@@ -1,10 +1,9 @@
-#![allow(clippy::useless_attribute)] //issue #2910
+// edition:2018

-#[macro_use]
-extern crate serde_derive;
+use serde::Deserialize;

 /// Tests that we do not lint for unused underscores in a `MacroAttribute`
 /// expansion
 #[deny(clippy::used_underscore_binding)]
 #[derive(Deserialize)]
 struct MacroAttributesTest {
```

Long story: Clippy tries to avoid this problem by adding `--extern` flags for crates that are also in the sysroot. In the rustc repo, two different `deps` dirs are passed (with `-L`) when compiling the test, `build/x86_64-unknown-linux-gnu/stage1-tools/release/deps` and `build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps`, but Clippy is only checking the last one. Moreover, only `rlibs` are taken into account, and `serde_derive` is a `.so`. Checking both dirs is also not an option, because in that case two instances of `quote` are found.

This is a long-standing issue in Clippy, which would be fixed by knowing the exact libraries cargo used when building `clippy_lints`. I will try to see what I can do to solve the general problem.

**Comment by Dylan-DPC** [comment]
@oli-obk this is ready for review
